packages/flutter/lib/src/material/app.dart

```diff
@@ -560,9 +376,13 @@ class _MaterialAppState extends State<MaterialApp> {
         key: GlobalObjectKey(this),
         navigatorKey: widget.navigatorKey,
         navigatorObservers: _haveNavigator ? _navigatorObservers : null,
+        pageRouteBuilder: <T>(RouteSettings settings, WidgetBuilder builder) =>
+          MaterialPageRoute<T>(settings: settings, builder: builder),
+        home: widget.home,
+        routes: widget.routes,
         initialRoute: widget.initialRoute,
-        onGenerateRoute: _haveNavigator ? _onGenerateRoute : null,
```

Comment:
A bit unfortunate that both CupertinoApp and MaterialApp maintain _haveNavigator just for the navigator observers. Looks like we can remove them and just let WidgetsApp deal with it can we not?

Reply:
Refactored to not store that state - just conditionally set the `_navigatorObservers` based on what was previously setting `_haveNavigator`